core/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala

```diff
@@ -91,6 +92,66 @@ private[spark] abstract class YarnSchedulerBackend(
   }
 
   /**
+   * Override the DriverEndpoint to add extra logic for the case when an executor is disconnected.
+   * We should check the cluster manager and find if the loss of the executor was caused by YARN
+   * force killing it due to preemption.
+   */
+  private class YarnDriverEndpoint(rpcEnv: RpcEnv, sparkProperties: ArrayBuffer[(String, String)])
+      extends DriverEndpoint(rpcEnv, sparkProperties) {
+
+    private val pendingDisconnectedExecutors = new HashSet[String]
+    private val handleDisconnectedExecutorThreadPool =
+      ThreadUtils.newDaemonCachedThreadPool("yarn-driver-handle-lost-executor-thread-pool")
+
+    /**
+     * When onDisconnected is received at the driver endpoint, the superclass DriverEndpoint
+     * handles it by assuming the Executor was lost for a bad reason and removes the executor
+     * immediately.
+     *
+     * In YARN's case however it is crucial to talk to the application master and ask why the
+     * executor had exited. In particular, the executor may have exited due to the executor
+     * having been preempted. If the executor "exited normally" according to the application
+     * master then we pass that information down to the TaskSetManager to inform the
+     * TaskSetManager that tasks on that lost executor should not count towards a job failure.
+     */
+    override def onDisconnected(rpcAddress: RpcAddress): Unit = {
```

Comment:
I guess the scary part is that if the executor died from an actual failure and tasks run on that bad executor, then we get more tasks that are marked as failed than we would have otherwise.